docs/design-archive/koordlet-runtime-hooks-mode.md

```diff
@@ -0,0 +1,87 @@
+# Design of Runtime Hook Running Mode in Koordlet
+
+## Summary
+`Runtime Hooks` includes a set of plugins which are responsible for the injections of resource isolation parameters 
+by pod attribute. When `Koord Runtime Proxy` running as a CRI Proxy, `Runtime Hooks` acts as the backend server. The
+mechanism of CRI Proxy can ensure the consistency of resource parameters during pod lifecycle. However, 
+`Koord Runtime Proxy` can only hijack CRI requests from kubelet for pods, the consistency of resource parameters in
+QoS class directory cannot be guaranteed. Besides, modification of pod parameters from third-party(e.g. manually) will
+also break the correctness of hook plugins.
+
+Therefore, a standalone running mode with reconciler for `Runtime Hooks` is necessary. Under `Standalone` running
+mode, resource isolation parameters will be injected asynchronously, keeping eventual consistency of the injected
+parameters for pod and QoS class even without `Runtime Hook Manager`.
+
+## Goals
+- `Runtime Hooks` Framework design for multiple running mode support.
+- Detailed design of `standalone` running mode.
+
+## None-Goals
+- Implements guide of runtime hook plugins. 
+
+## Architecture
+![image](../images/koordlet-hooks-running-mode.svg)
+
+## Framework
+There are two running mode in `Runtime Hooks` working at the same time: `proxy` and `standalone`. `Proxy Server` module
+is the ingress of `proxy` running mode, receiving Runtime Hook protocol during pod lifecycle. `Reconciler` is for
+`standalone` running mode; it will receive events when pod related cgroups files updated. Both module will
+convert its requests to internal requests for hook plugins and parse the injections parameters from internal response.
+For `proxy` running mode, injections will be wrote back to `Runtime Hook` response, and for `standalone` running mode,
+injections will be applied on OS directly.
+
+The internal protocol adapts two running mode inside the hook plugin framework, so that plugin developers does not need 
+to rewrite hook functions of pod for different working mode, using the same functions to register. `Reconciler` takes
+resource level(QoS class, pod and container) and filename as the input for the maintenance of cgroup resources, and can
+be easily extended for more kinds of resources in the future.
```

Comment:
Do plugin developers need to know this difference of return values?

Reply:
plugin developers need to:
register config parse function for policy(NodeSLO);
register hook function for proxy interception;
register reconcile function for final consistency.

`hook function` and `reconcile function` using same internal protocol, so developers can use same implementation, do not need to know the difference.